### ISSUE TITLE
Pyright hangs on some recursive aliases

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#9497]
**Describe the bug**

Pyright hangs in certain cases involving recursive type aliases.

**Code or Screenshots**

```python
from typing import TypeVar, Generic

U = TypeVar("U")
T = TypeVar("T")

class A(Generic[T]):
    pass

class B(Generic[T]):
    pass

class Tagged(Generic[T]):
    pass

NoTag = A["Either[U]"] | B["Either[U]"]
Either = NoTag[U] | Tagged[NoTag[U]]

### If this is moved to a separate file, there is no problem
Problem = Either[U]
```

Here's my pyrightconfig.json:

```json
{
  "pythonVersion": "3.11",
  "typeCheckingMode": "strict",
}
```

Changing this to pythonVersion 3.12 and modifying the source accordingly fixes the problem:

```python
type NoTag[U] = A[Either[U]] | B[Either[U]]
type Either[U] = NoTag[U] | Tagged[NoTag[U]]
type Problem[U] = Either[U]
```

With Python 3.11, using the `: TypeAlias` annotation does not fix the problem.

**VS Code extension or command-line**

```sh
$ ./node_modules/.bin/pyright --version
pyright 1.1.389
```

This issue also occurs in pyright 1.1.385.

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem, and I'll investigate further.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.390.
